File: duck_vlm/intent.py
"""Free-form task text -> coarse intent, so plugins can match semantics rather
than requiring an exact task id.

The curated scene tasks stay authoritative when their id is known; this module
only decides what to do with a user sentence that matched nothing.
"""
from __future__ import annotations

# Ordered by specificity: the first family that matches wins, so "turn right and
# stop" is a pose task and not an approach task just because it says "then stop".
RECOVER = ("recover", "get up", "stand up", "get back up", "on its feet",
           "起身", "站起来", "恢复", "爬起来")
ORBIT = ("orbit", "circle around", "go behind", "side-rear", "walk around",
         "绕", "侧后方", "转到后面")
GAIT = ("rough ground", "uneven", "gravel", "corridor", "narrow", "ramp",
        "不平", "碎石", "走廊", "窄道", "斜坡")
# Note: no bare "stop"/"停止" here. "go to the cube and stop next to it" is an approach
# task whose final condition happens to be stopping, and classifying it as a pose task
# would switch exploration off.
SPIN_NOTE = "spin"  # a bare spin is a rotation request, not an orbit
POSE = ("turn right", "turn left", "turn around", "spin in place", "spin around", "spin", "stand still", "sit",
        "stay there", "wait there", "原地", "转身", "站着", "别动", "右转", "左转")
MANIPULATE = ("kick", "push", "shove", "nudge", "bring", "retrieve", "carry", "deliver",
              "take the", "take it", "put it", "pick", "grab", "move it",
              "put it in", "into the",
              "踢", "推", "捅", "送到", "运到", "带到", "拿给", "放进", "拿", "捡", "抓")
APPROACH = ("go to", "go over", "walk to", "walk towards", "approach", "reach", "find",
            "get to", "move to", "head to", "come here", "come over", "come and",
            "过去", "走向", "走到", "靠近", "找到", "去找", "过来", "来这里",
            "去", "到", "找", "看看")

# Negation: "don't go near the blue cube" must not resolve the blue cube as a goal.
NEGATION = ("don't", "do not", "avoid", "stay away", "keep away", "not go",
            "别", "不要", "不准", "远离", "避免")


def resolve_intent(task_id: str, task_text: str) -> str:
    """One of recover | orbit | gait | pose | manipulate | approach | unknown."""
    text = (task_text or "").lower()
    if any(w in text for w in NEGATION):
        return "avoid"
    # Manipulation is checked before orbit/gait: "push the ball around the pillar into
    # the zone" is a manipulation task that merely routes around an obstacle.
    for name, words in (("recover", RECOVER), ("manipulate", MANIPULATE), ("orbit", ORBIT),
                        ("gait", GAIT), ("pose", POSE), ("approach", APPROACH)):
        if any(w in text for w in words):
            return name
    return "unknown"


def is_blocking_intent(task_id: str, task_text: str) -> bool:
    """True when walking to a waypoint would fight the objective."""
    return resolve_intent(task_id, task_text) in ("recover", "orbit", "gait", "pose")


def pick_target(hits, task_text: str) -> tuple[str, str | None]:
    """Choose which body the sensor should track, from scanned entity hits.

    ``hits`` is what SceneSpec.scan_text returns: [(label, body, kind), ...] with the
    most specific label first. For "kick the ball into the green zone" the longest
    match is the zone, but the thing to act on is the ball, so manipulation intents
    prefer a non-zone entity. Returns (body, matched label or None).
    """
    if not hits:
        return "ball", None
    intent = resolve_intent("", task_text)
    if intent == "avoid":
        # do not track (and therefore chase) the thing the user asked us to avoid
        return "ball", None
    if intent == "manipulate":
        for label, body, kind in hits:
            if kind != "zone":
                return body, label
    label, body, _kind = hits[0]
    return body, label

File: web_sim/src/intent.js
/**
 * 自由文本 → 粗粒度意图 —— `duck_vlm/intent.py` 的移植。
 *
 * 用途：场景包里那些精选任务（task id）仍然权威；但当用户随手写一句话时，
 * 我们需要知道他说的是「走过去」还是「踢到某处」，才能决定：
 *   - 跟踪哪个目标（踢球时要盯球，而不是盯绿色区域）；
 *   - 要不要开启探索（原地转圈类任务就别乱走）。
 * 词表顺序就是优先级，先匹配到的家族胜出。
 */

export const RECOVER = ["recover", "get up", "stand up", "get back up", "on its feet",
                        "起身", "站起来", "恢复", "爬起来"];
export const ORBIT = ["orbit", "circle around", "go behind", "side-rear", "walk around",
                      "绕", "侧后方", "转到后面"];
export const GAIT = ["rough ground", "uneven", "gravel", "corridor", "narrow", "ramp",
                     "不平", "碎石", "走廊", "窄道", "斜坡"];
// 注意这里**没有**光秃秃的 stop：『走到方块旁边然后停下』是接近任务，
// 归成姿态任务会让探索插件被关掉。
export const POSE = ["turn right", "turn left", "turn around", "spin in place", "spin around", "spin",
                     "stand still", "sit", "stay there", "wait there",
                     "原地", "转身", "站着", "别动", "右转", "左转"];
export const MANIPULATE = ["kick", "push", "shove", "nudge", "bring", "retrieve", "carry", "deliver",
                           "take the", "take it", "put it", "pick", "grab", "move it",
                           "put it in", "into the",
                           "踢", "推", "捅", "送到", "运到", "带到", "拿给", "放进", "拿", "捡", "抓"];
export const APPROACH = ["go to", "go over", "walk to", "walk towards", "approach", "reach", "find",
                         "get to", "move to", "head to", "come here", "come over", "come and",
                         "过去", "走向", "走到", "靠近", "找到", "去找", "过来", "来这里",
                         "去", "到", "找", "看看"];
// 否定：「别靠近蓝方块」不能把蓝方块当成目标。
export const NEGATION = ["don't", "do not", "avoid", "stay away", "keep away", "not go",
                         "别", "不要", "不准", "远离", "避免"];

/** @returns {"recover"|"orbit"|"gait"|"pose"|"manipulate"|"approach"|"avoid"|"unknown"} */
export function resolveIntent(taskId, taskText) {
  const text = String(taskText || "").toLowerCase();
  if (NEGATION.some((w) => text.includes(w))) return "avoid";
  // 操作类排在绕行/地形之前：『把球绕着柱子推进区域』是操作任务，只是在避障。
  for (const [name, words] of [["recover", RECOVER], ["manipulate", MANIPULATE], ["orbit", ORBIT],
                               ["gait", GAIT], ["pose", POSE], ["approach", APPROACH]]) {
    if (words.some((w) => text.includes(w))) return name;
  }
  return "unknown";
}

/** 走向航点会和目标打架的意图（此时不该开探索/寻路）。 */
export function isBlockingIntent(taskId, taskText) {
  return ["recover", "orbit", "gait", "pose"].includes(resolveIntent(taskId, taskText));
}

/**
 * 从「句子里提到的实体」里挑出该被跟踪的 body。
 * hits 来自 SceneIndex.scanText，最具体的标签在前。
 * 「把球踢进绿色区域」里最长匹配是区域，但要操作的是球 → 操作类意图优先非 zone 实体。
 */
export function pickTarget(hits, taskText) {
  if (!hits || !hits.length) return { body: "ball", label: null };
  const intent = resolveIntent("", taskText);
  if (intent === "avoid") return { body: "ball", label: null };   // 别去追用户让你躲开的东西
  if (intent === "manipulate") {
    for (const [label, body, kind] of hits) if (kind !== "zone") return { body, label };
  }
  const [label, body] = hits[0];
  return { body, label };
}
